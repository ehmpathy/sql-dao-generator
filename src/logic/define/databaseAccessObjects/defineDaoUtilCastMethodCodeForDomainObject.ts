import { camelCase, snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyMetadata,
  DomainObjectPropertyType,
  DomainObjectReferenceMetadata,
} from 'domain-objects-metadata';
import { isPresent } from 'simple-type-guards';

import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { defineOutputTypeOfFoundDomainObject } from './defineOutputTypeOfFoundDomainObject';

export const defineDaoUtilCastMethodCodeForDomainObject = ({
  domainObject,
  sqlSchemaRelationship,
}: {
  domainObject: DomainObjectMetadata;
  sqlSchemaRelationship: SqlSchemaToDomainObjectRelationship;
}) => {
  // define some useful constants
  const hasUuidProperty = !!domainObject.properties.uuid;

  // define the referenced domain objects to hydrate
  const nestedDomainObjectNames = Object.values(domainObject.properties)
    .map((property) => {
      if (property.type === DomainObjectPropertyType.REFERENCE)
        return (property.of as DomainObjectReferenceMetadata).name;
      if (
        property.type === DomainObjectPropertyType.ARRAY &&
        (property.of as DomainObjectPropertyMetadata).type === DomainObjectPropertyType.REFERENCE
      )
        return ((property.of as DomainObjectPropertyMetadata).of as DomainObjectReferenceMetadata).name;
      return null;
    })
    .filter(isPresent)
    .sort();

  // define the imports
  const imports = [
    // always present imports
    `import { HasId${hasUuidProperty ? ', HasUuid' : ''} } from 'simple-type-guards';`,
    '', // split module from relative imports
    `import { ${[domainObject.name, ...nestedDomainObjectNames].sort().join(', ')} } from '$PATH_TO_DOMAIN_OBJECT';`, // import this domain object; note: higher level function will swap out the import path
    `import { SqlQueryFind${domainObject.name}ByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';`,
  ];

  // define the output type
  const outputType = defineOutputTypeOfFoundDomainObject(domainObject);

  const propertiesToInstantiate = [
    'id: dbObject.id',
    hasUuidProperty ? 'uuid: dbObject.uuid' : null,
    ...sqlSchemaRelationship.properties.map(({ sqlSchema: sqlSchemaProperty, domainObject: domainObjectProperty }) => {
      // enum case
      if (domainObjectProperty.type === DomainObjectPropertyType.ENUM)
        return `${domainObjectProperty.name}: dbObject.${sqlSchemaProperty.name} as ${domainObject.name}['${domainObjectProperty.name}']`;

      // non-reference case
      if (!sqlSchemaProperty.reference) return `${domainObjectProperty.name}: dbObject.${sqlSchemaProperty.name}`;

      // referenced by uuid case
      if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID) {
        // solo reference case
        if (!sqlSchemaProperty.isArray)
          return `${domainObjectProperty.name}: dbObject.${snakeCase(domainObjectProperty.name)}`;

        // array reference case
        return `${domainObjectProperty.name}: dbObject.${snakeCase(domainObjectProperty.name)} as string[]`; // as string array since we have an array of uuids - but the type defs generated from sql will complain that it could be string[] or number[] or null (not smart enough to look all the way through fn defs yet)
      }

      // directly nested case
      if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) {
        // solo reference case
        if (!sqlSchemaProperty.isArray)
          return `${domainObjectProperty.name}: new ${sqlSchemaProperty.reference.of.name}(dbObject.${snakeCase(
            domainObjectProperty.name,
          )} as ${sqlSchemaProperty.reference.of.name})`;

        // array reference case
        return `${domainObjectProperty.name}: dbObject.${snakeCase(domainObjectProperty.name)}.map((${camelCase(
          sqlSchemaProperty.reference.of.name,
        )}) => new ${sqlSchemaProperty.reference.of.name}(${camelCase(sqlSchemaProperty.reference.of.name)} as ${
          sqlSchemaProperty.reference.of.name
        }))`;
      }

      // handle unexpected case (each case should have been handled above)
      throw new Error('unexpected property type. this is a bug within sql-dao-generator'); // fail fast if reached here
    }),
  ].filter(isPresent);

  // define the content
  const code = `
${imports.join('\n')}

export const castFromDatabaseObject = ({
  dbObject,
}: {
  dbObject: SqlQueryFind${domainObject.name}ByIdOutput;
}): ${outputType} =>
  new ${domainObject.name}({
    ${propertiesToInstantiate.join(',\n    ')},
  }) as ${outputType};
`.trim();

  // return the code
  return code;
};
