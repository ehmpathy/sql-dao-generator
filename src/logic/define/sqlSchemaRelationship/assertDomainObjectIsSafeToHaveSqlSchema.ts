import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { UserInputError } from '../../UserInputError';

export const assertDomainObjectIsSafeToHaveSqlSchema = ({ domainObject }: { domainObject: DomainObjectMetadata }) => {
  // make sure that entities have "unique" and "updatable" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_ENTITY) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason:
          "domain entities must have at least one 'unique' property defined in order for a schema to be generated.",
        domainObjectName: domainObject.name,
        potentialSolution:
          "(note: if its not unique on any natural keys, please specify the 'uuid' as it's only unique property. e.g., `['uuid']`) ",
      });
    if (!Array.isArray(domainObject.decorations.updatable))
      throw new UserInputError({
        reason:
          "domain entities must have their 'updatable' properties defined in order for a schema to be generated. ",
        domainObjectName: domainObject.name,
        potentialSolution:
          '(note: if it has no updatable properties, please specify this with an empty array. e.g., `[]`) ',
      });
  }

  // make sure that value objects do not have unique or updatable defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) {
    if (domainObject.decorations.unique)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have their 'unique' properties specified. value objects are unique on all of their properties by definition.",
        domainObjectName: domainObject.name,
      });
    if (domainObject.decorations.updatable)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have any 'updatable' properties specified. value objects do not have any updatable properties by definition.",
        domainObjectName: domainObject.name,
      });
  }

  // make sure that events have "unique" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_EVENT) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason: "domain events must have at least one 'unique' property defined in order for a schema to be generated",
        domainObjectName: domainObject.name,
      });
  }
};
