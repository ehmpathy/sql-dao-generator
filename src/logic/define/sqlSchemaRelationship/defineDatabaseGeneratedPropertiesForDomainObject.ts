import { DomainObjectMetadata } from 'domain-objects-metadata';
import { isPresent } from 'type-fns';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';

export const defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject = ({
  domainObject,
}: {
  domainObject: DomainObjectMetadata;
}) => {
  const isUniqueOnUuid = domainObject.decorations.unique?.includes('uuid');
  const hasUpdatableProperties =
    (domainObject.decorations.updatable?.length ?? 0) > 0;

  // define which sql schema properties are generated by the db
  return [
    new SqlSchemaPropertyMetadata({
      name: 'id',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    !isUniqueOnUuid
      ? new SqlSchemaPropertyMetadata({
          name: 'uuid',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // "uuid" is only autogenerated if the domain object is not unique on it
    new SqlSchemaPropertyMetadata({
      name: 'created_at',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'effective_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "effective_at" defined if they're updatable
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'updated_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "updated_at" defined if they're updatable
  ].filter(isPresent);
};
