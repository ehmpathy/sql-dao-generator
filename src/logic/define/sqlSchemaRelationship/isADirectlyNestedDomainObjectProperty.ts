import {
  DomainObjectPropertyMetadata,
  isDomainObjectArrayProperty,
  isDomainObjectReferenceProperty,
} from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain';
import { isAUserDefinedDomainObjectProperty } from './isAUserDefinedDomainObjectProperty';

/**
 * specifies whether this domain object property is a reference to another domain object directly by nesting (in array or solo)
 */
export const isADirectlyNestedDomainObjectProperty = (propertyRelationship: {
  domainObject: DomainObjectPropertyMetadata | null;
  sqlSchema: SqlSchemaPropertyMetadata;
}): propertyRelationship is {
  domainObject: DomainObjectPropertyMetadata;
  sqlSchema: SqlSchemaPropertyMetadata;
} => {
  if (!isAUserDefinedDomainObjectProperty(propertyRelationship)) return false; // if not user defined, then its def not a reference
  if (isDomainObjectReferenceProperty(propertyRelationship.domainObject))
    return true; // if its directly nested, its directly nested
  if (
    isDomainObjectArrayProperty(propertyRelationship.domainObject) &&
    isDomainObjectReferenceProperty(propertyRelationship.domainObject.of)
  )
    return true; // if its an array of directly nested objects, still true
  return false; // otherwise, false
};
