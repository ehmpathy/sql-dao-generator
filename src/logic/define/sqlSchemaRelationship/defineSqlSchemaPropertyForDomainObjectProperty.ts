import { camelCase, snakeCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectPropertyMetadata, DomainObjectPropertyType } from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaReferenceForDomainObjectProperty } from './defineSqlSchemaReferenceForDomainObjectProperty';

export const defineSqlSchemaPropertyForDomainObjectProperty = ({
  property,
  domainObject,
  allDomainObjects,
}: {
  property: DomainObjectPropertyMetadata;
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaPropertyMetadata => {
  // sanity check that property name is in camel case
  if (camelCase(property.name) !== property.name)
    throw new UserInputError({
      reason: 'property names must be camel case',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // sanity check that this is not called for a database generated property (since that should be handled elsewhere)
  const databaseGeneratedProperties = defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({ domainObject });
  if (databaseGeneratedProperties.map(({ name }) => name).includes(snakeCase(property.name)))
    throw new UnexpectedCodePathDetectedError({
      reason: '`defineSqlSchemaPropertyForDomainObjectProperty was called for an db-generated property',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // define data about this property
  const isUpdatable = domainObject.decorations.updatable?.includes(property.name) ?? false;
  const isNullable = !!property.nullable;
  const isArray = property.type === DomainObjectPropertyType.ARRAY;

  // determine if this is a reference of another domain object
  const reference = defineSqlSchemaReferenceForDomainObjectProperty({
    property,
    domainObject,
    allDomainObjects,
  });

  // define the schema property name
  const schemaPropertyName = (() => {
    const baseSchemaPropertyName = snakeCase(property.name);
    if (reference) {
      if (isArray) {
        if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
          return `${baseSchemaPropertyName.replace(/s$/, '')}_ids`; // suffix w/ ids, since its an array of fks
        if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
          return `${baseSchemaPropertyName.replace(/_uuids$/, '')}_ids`; // suffix w/ ids, since its an array of fks
      }
      if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) return `${baseSchemaPropertyName}_id`; // suffix w/ id, since its a fk
      if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
        return `${baseSchemaPropertyName.replace(/_uuid$/, '')}_id`; // suffix w/ id, since its a fk
    }
    return baseSchemaPropertyName; // otherwise, its just the regular name, no modifications
  })();

  return new SqlSchemaPropertyMetadata({
    name: schemaPropertyName,
    isUpdatable,
    isNullable,
    isArray,
    isDatabaseGenerated: false, // this method is only called for properties that are not database generated (as checked above)
    reference,
  });
};
