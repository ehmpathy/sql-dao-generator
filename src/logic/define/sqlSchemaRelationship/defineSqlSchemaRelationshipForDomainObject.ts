import { snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';
import { isNotADatabaseGeneratedProperty } from './isNotADatabaseGeneratedProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaToDomainObjectRelationship => {
  // figure out some relevant info
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const databaseGeneratedSqlSchemaProperties =
    defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({
      domainObject,
    });
  const propertiesRelationship = [
    /// autogenerated values
    ...databaseGeneratedSqlSchemaProperties.map((sqlSchemaProperty) => {
      // see if this property was defined on the domain object (i.e., they want the value for it)
      const domainObjectProperty =
        Object.values(domainObject.properties).find(
          (property) => snakeCase(property.name) === sqlSchemaProperty.name,
        ) ?? null;

      // if so, validate that it has the correct type
      if (domainObjectProperty) {
        const expectedType = (() => {
          if (sqlSchemaProperty.name === 'id')
            return DomainObjectPropertyType.NUMBER;
          if (sqlSchemaProperty.name === 'uuid')
            return DomainObjectPropertyType.STRING;
          if (sqlSchemaProperty.name === 'created_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'effective_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'updated_at')
            return DomainObjectPropertyType.DATE;
          throw new UnexpectedCodePathDetectedError({
            reason:
              'unexpected database generated property name. this is a bug within sql-schema-generator.', // fail fast if expectations not met
            domainObjectName: domainObject.name,
            domainObjectPropertyName: sqlSchemaProperty.name,
          });
        })();
        if (domainObjectProperty.type !== expectedType)
          throw new UserInputError({
            reason: `reserved database-generated property does not match expected type. expected ${expectedType} but received ${domainObjectProperty.type}`,
            domainObjectName: domainObject.name,
            domainObjectPropertyName: domainObjectProperty.name,
          });
      }

      // return the relationship
      return {
        sqlSchema: sqlSchemaProperty,
        domainObject: domainObjectProperty,
      };
    }),

    // non autogenerated values
    ...Object.values(domainObject.properties)
      .filter(
        (domainObjectProperty) =>
          !databaseGeneratedSqlSchemaProperties
            .map(({ name }) => name)
            .includes(snakeCase(domainObjectProperty.name)), // filter out the db generated values, because we already got them above
      )
      .map((domainObjectProperty) => {
        return {
          sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({
            property: domainObjectProperty,
            domainObject,
            allDomainObjects,
          }),
          domainObject: domainObjectProperty,
        };
      }),
  ];

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_LITERAL
        ? propertiesRelationship // if its a literal, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
            .filter(isNotADatabaseGeneratedProperty)
            .map(
              ({ domainObject: domainObjectProperty }) =>
                domainObjectProperty.name,
            )
        : domainObject.decorations.unique; // otherwise, its unique on what was specified

    // convert the domain object prop names into sql schema prop names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) =>
            propertyRelationship.domainObject?.name ===
            domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new UserInputError({
            reason: 'Unique keys must be properties of the domain object.',
            domainObjectName: domainObject.name,
            potentialSolution: ` Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'? (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          });
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      alias: {
        domainObject: domainObject.decorations.alias,
      },
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
