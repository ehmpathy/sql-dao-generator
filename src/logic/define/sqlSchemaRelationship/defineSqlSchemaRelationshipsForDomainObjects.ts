import { DomainObjectMetadata } from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from './defineSqlSchemaRelationshipForDomainObject';

/**
 * uses the full list of domain objects to define a relationship between its DomainObject representation and its SqlSchema representation
 */
export const defineSqlSchemaRelationshipsForDomainObjects = ({
  domainObjects,
}: {
  domainObjects: DomainObjectMetadata[];
}) =>
  domainObjects.map((domainObject) =>
    defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: domainObjects,
    }),
  );
