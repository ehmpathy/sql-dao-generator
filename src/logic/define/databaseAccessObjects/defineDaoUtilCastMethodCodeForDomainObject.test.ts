import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from '../sqlSchemaRelationship/defineSqlSchemaRelationshipForDomainObject';
import { defineDaoUtilCastMethodCodeForDomainObject } from './defineDaoUtilCastMethodCodeForDomainObject';

describe('defineDaoUtilCastMethodCodeForDomainObject', () => {
  it('should look correct for simple literal', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Geocode',
      extends: DomainObjectVariant.DOMAIN_LITERAL,
      properties: {
        id: { name: 'id', type: DomainObjectPropertyType.NUMBER },
        latitude: { name: 'latitude', type: DomainObjectPropertyType.NUMBER },
        longitude: { name: 'longitude', type: DomainObjectPropertyType.NUMBER },
      },
      decorations: {
        unique: null,
        updatable: null,
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineDaoUtilCastMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
    });

    // log an example
    expect(code).toContain("import { Geocode } from '$PATH_TO_DOMAIN_OBJECT'");
    expect(code).toContain(
      "import { SqlQueryFindGeocodeByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';",
    );
    expect(code).toContain('dbObject: SqlQueryFindGeocodeByIdOutput');
    expect(code).toContain('new Geocode({');
    expect(code).toMatchSnapshot();
  });
  it('should look correct for simple domain entity', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Carriage',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        cin: {
          name: 'cin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        carries: {
          name: 'carries',
          type: DomainObjectPropertyType.ENUM,
          of: ['PASSENGER', 'FREIGHT'],
          required: true,
        },
        capacity: {
          name: 'capacity',
          type: DomainObjectPropertyType.NUMBER,
          nullable: true,
        },
      },
      decorations: {
        unique: ['cin'],
        updatable: ['capacity'],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineDaoUtilCastMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
    });

    // log an example
    expect(code).toContain("import { Carriage } from '$PATH_TO_DOMAIN_OBJECT'");
    expect(code).toContain(
      "import { SqlQueryFindCarriageByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';",
    );
    expect(code).toContain('dbObject: SqlQueryFindCarriageByIdOutput');
    expect(code).toContain('new Carriage({');
    expect(code).toMatchSnapshot();
  });
  it('should look correct for a domain event with a static referenced array', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'TrainLocatedEvent',
      extends: DomainObjectVariant.DOMAIN_EVENT,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        trainUuid: {
          name: 'trainUuid',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        occurredAt: {
          name: 'occurredAt',
          type: DomainObjectPropertyType.DATE,
          required: true,
        },
        geocodes: {
          name: 'geocodes',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
          required: true,
        },
      },
      decorations: {
        unique: ['trainUuid', 'occurredAt'],
        updatable: [],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [] as DomainObjectMetadata[],
    });

    // run it
    const code = defineDaoUtilCastMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
    });

    // log an example
    expect(code).toContain(
      "import { TrainLocatedEvent } from '$PATH_TO_DOMAIN_OBJECT'",
    );
    expect(code).toContain(
      "import { SqlQueryFindGeocodeByIdOutput, SqlQueryFindTrainLocatedEventByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';",
    );
    expect(code).toContain(
      "import { castFromDatabaseObject as castGeocodeFromDatabaseObject } from '../geocodeDao/castFromDatabaseObject';",
    );
    expect(code).toContain('dbObject: SqlQueryFindTrainLocatedEventByIdOutput');
    expect(code).toContain('new TrainLocatedEvent({');
    expect(code).toContain(
      'geocodes: (dbObject.geocodes as SqlQueryFindGeocodeByIdOutput[]).map(castGeocodeFromDatabaseObject)',
    ); // instantiate nested array of geocodes
    expect(code).toMatchSnapshot();
  });
  it('should look correct for domain entity with references, array and solo, implicit and direct', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Train',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        tin: {
          name: 'tin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        homeStationGeocode: {
          name: 'homeStationGeocode',
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_LITERAL,
          },
        },
        leadEngineerUuid: {
          name: 'leadEngineerUuid',
          type: DomainObjectPropertyType.STRING,
        },
        badges: {
          name: 'badges',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'TrainBadge',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        locomotiveUuids: {
          name: 'locomotiveUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.STRING,
          },
        },
      },
      decorations: {
        unique: ['tin'],
        updatable: [
          'homeStationGeocode',
          'badges',
          'locomotiveUuids',
          'leadEngineerUuid',
        ],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [
        domainObject,
        { name: 'Locomotive', extends: DomainObjectVariant.DOMAIN_ENTITY },
        { name: 'TrainEngineer', extends: DomainObjectVariant.DOMAIN_ENTITY },
      ] as DomainObjectMetadata[],
    });

    // run it
    const code = defineDaoUtilCastMethodCodeForDomainObject({
      domainObject,
      sqlSchemaRelationship,
    });

    // log an example
    expect(code).toContain("import { Train } from '$PATH_TO_DOMAIN_OBJECT'");
    expect(code).toContain(
      "import { SqlQueryFindGeocodeByIdOutput, SqlQueryFindTrainBadgeByIdOutput, SqlQueryFindTrainByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';",
    );
    expect(code).toContain(
      "import { castFromDatabaseObject as castGeocodeFromDatabaseObject } from '../geocodeDao/castFromDatabaseObject';",
    );
    expect(code).toContain(
      "import { castFromDatabaseObject as castTrainBadgeFromDatabaseObject } from '../trainBadgeDao/castFromDatabaseObject';",
    );
    expect(code).toContain('dbObject: SqlQueryFindTrainByIdOutput');
    expect(code).toContain('new Train({');
    expect(code).toContain(
      'homeStationGeocode: castGeocodeFromDatabaseObject(dbObject.home_station_geocode as SqlQueryFindGeocodeByIdOutput)',
    ); // instantiate nested array of geocodes
    expect(code).toContain('leadEngineerUuid: dbObject.lead_engineer_uuid');
    expect(code).toContain(
      'badges: (dbObject.badges as SqlQueryFindTrainBadgeByIdOutput[]).map(castTrainBadgeFromDatabaseObject)',
    ); // instantiate nested array of geocodes
    expect(code).toContain(
      'locomotiveUuids: dbObject.locomotive_uuids as string[]',
    );
    expect(code).toMatchSnapshot();
  });
});
