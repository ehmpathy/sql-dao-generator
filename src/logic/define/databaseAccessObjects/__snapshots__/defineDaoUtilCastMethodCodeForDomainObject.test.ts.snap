// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineDaoUtilCastMethodCodeForDomainObject should look correct for a domain event with a static referenced array 1`] = `
"import { HasId } from 'simple-type-guards';

import { Geocode, TrainLocatedEvent } from '$PATH_TO_DOMAIN_OBJECT';
import { SqlQueryFindTrainLocatedEventByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';

export const castFromDatabaseObject = ({
  dbObject,
}: {
  dbObject: SqlQueryFindTrainLocatedEventByIdOutput;
}): HasId<TrainLocatedEvent> =>
  new TrainLocatedEvent({
    id: dbObject.id,
    trainUuid: dbObject.train_uuid,
    occurredAt: dbObject.occurred_at,
    geocodes: (dbObject.geocodes as Geocode[]).map((geocode) => new Geocode(geocode)),
  }) as HasId<TrainLocatedEvent>;"
`;

exports[`defineDaoUtilCastMethodCodeForDomainObject should look correct for domain entity with references, array and solo, implicit and direct 1`] = `
"import { HasId, HasUuid } from 'simple-type-guards';

import { Geocode, Train, TrainBadge } from '$PATH_TO_DOMAIN_OBJECT';
import { SqlQueryFindTrainByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';

export const castFromDatabaseObject = ({
  dbObject,
}: {
  dbObject: SqlQueryFindTrainByIdOutput;
}): HasId<HasUuid<Train>> =>
  new Train({
    id: dbObject.id,
    uuid: dbObject.uuid,
    tin: dbObject.tin,
    homeStationGeocode: new Geocode(dbObject.home_station_geocode as Geocode),
    leadEngineerUuid: dbObject.lead_engineer_uuid,
    badges: (dbObject.badges as TrainBadge[]).map((trainBadge) => new TrainBadge(trainBadge)),
    locomotiveUuids: dbObject.locomotive_uuids as string[],
  }) as HasId<HasUuid<Train>>;"
`;

exports[`defineDaoUtilCastMethodCodeForDomainObject should look correct for simple domain entity 1`] = `
"import { HasId, HasUuid } from 'simple-type-guards';

import { Carriage } from '$PATH_TO_DOMAIN_OBJECT';
import { SqlQueryFindCarriageByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';

export const castFromDatabaseObject = ({
  dbObject,
}: {
  dbObject: SqlQueryFindCarriageByIdOutput;
}): HasId<HasUuid<Carriage>> =>
  new Carriage({
    id: dbObject.id,
    uuid: dbObject.uuid,
    cin: dbObject.cin,
    carries: dbObject.carries as Carriage['carries'],
    capacity: dbObject.capacity,
  }) as HasId<HasUuid<Carriage>>;"
`;

exports[`defineDaoUtilCastMethodCodeForDomainObject should look correct for simple value object 1`] = `
"import { HasId } from 'simple-type-guards';

import { Geocode } from '$PATH_TO_DOMAIN_OBJECT';
import { SqlQueryFindGeocodeByIdOutput } from '$PATH_TO_GENERATED_SQL_TYPES';

export const castFromDatabaseObject = ({
  dbObject,
}: {
  dbObject: SqlQueryFindGeocodeByIdOutput;
}): HasId<Geocode> =>
  new Geocode({
    id: dbObject.id,
    latitude: dbObject.latitude,
    longitude: dbObject.longitude,
  }) as HasId<Geocode>;"
`;
